Distribution graph

mokaddem opened the pull request:
#### What does it do?
It displays a summarized view of how information contained in an event is distributed among different actors allowing to:
    - quickly verify that the distribution of all elements matches the intended distribution level
    - quickly verify distribution inconsistencies in the event (attributes being *all communities* with an event being *org only*)

#### New feature
- Add a popover next to the distribution level with:
    - a donut chart showing the repartition of the distribution for the event, the attribute and the object's attribute
    - a progressbar showing a summary of the event distribution
    - a progressbar showing the number of involved sharing group
    - additional information about distribution including all concerned parties (organization and connected server)
- Possibility to filter elements in the event based on their distribution

#### MISP core changes
- **[EventController.php]** ``__valueInFieldAttribute`` can be used with or (``|`` char)
- **[EventController.php]** ``viewEventAttributes`` can overwrite the default ``filterColumns`` in order to bypass ``validFilters``, so that we can filter on other fields (e.g. distribution).
- **[misp.js]** ``filterAttributes`` supports filtering with specified columns

#### New dependencies
- ``Chart.js`` 

#### Release Type:
- [ ] Major
- [X] Minor
- [X] Patch

# Picture
![image](https://user-images.githubusercontent.com/6977223/39577947-0920420a-4ee3-11e8-9e91-871233545e01.png)

mokaddem committed:
Initial version of the distribution graph

mokaddem committed:
Merge remote-tracking branch 'upstream/2.4' into distributionGraph

mokaddem committed:
Allow filtering attributes based on specific columns (previsouly not
accessible) like distribution.
Partial support of onClick for distribution graph.

mokaddem committed:
Possibility to filter valueInFieldAttribute with multiple value.
distribution graph support inherit distribution level

mokaddem committed:
fix: support of filtering for distribution=0 (empty(0) is true ini
php). Also, only consider attr and obj_attr (ignoring object as they
only carry meta-data)

mokaddem committed:
feature: progress bar showing the range of the maximum distribution of
all items. Moved radar graph and  progressbar in a popover.

mokaddem committed:
Updated ACLComponent

mokaddem committed:
removed useless codes

mokaddem committed:
Changed distribution graph popover title

mokaddem committed:
Changed behavior of distribution progressbar: Display event distribution
along with the maximum distribution level of the items inside the event.

mokaddem committed:
fix: replaced hardcoded eventID by the real event id

mokaddem committed:
center distribution graph inside the popover

mokaddem committed:
Changed distribution label in distribution graph (removed distribution
number)

mokaddem committed:
Merge remote-tracking branch 'upstream/2.4' into distributionGraph

mokaddem committed:
new: Possibility to view connected communities and concerned sharing groups in distribution graph's tooltip

mokaddem committed:
fix: avoid redrawin distribution graph when closin its popover + reset
pb ticks offset at each draw.

mokaddem committed:
fix: fixed distribution level swapping when filtering from the
distribution
chg: moved styling into css
new: Loading gif when building the distribution graph

mokaddem committed:
chg: removed useless prints

mokaddem committed:
chg: Changed distribution graph popover title

mokaddem committed:
new: show elements having a distribution lower than the event
distribution in the distribution graph

mokaddem committed:
chg: moved sharing group outside of the distribution progressbar (as it is a special case), distribution range is displayed when clicking on the pb labels and lots of minor improvements.

mokaddem committed:
chg: replaced radar chart to doughnut chart

mokaddem committed:
new: possibility to show/hide distribution repartition of
event/attr/objAttr
chg: layout adaptation

mokaddem committed:
chg: add additional distribution info about to whom we are sharing even
if we don't have element on this distribution level

mokaddem committed:
chg: sanitization of data for distribution graph

mokaddem committed:
chg: updated description tooltip text

mokaddem committed:
chg: doughnut part color

mokaddem committed:
new: added chartjs dependency

mokaddem committed:
chg: show all by default

mokaddem committed:
chg: update __query version

mokaddem committed:
chg: [distributionGraph] support of the sharing group event distribution
chg: [distributionGraph] code cleanup

mokaddem committed:
chg: [DistributionGraph] added ``distribution description`` text in the info
popup

mokaddem commented:
Feedback welcome! :)

mokaddem committed:
Merge remote-tracking branch 'upstream/2.4' into distributionGraphDonut

adulau commented:
As already discussed, you implemented the changes and it seems incredibility good. I just need a final go before merging from the Master Chief of the MVC @iglocska to be sure that the approach of getting list of sync servers, orgs and sharing groups is fine. Thank you!

mokaddem committed:
chg: Trying not to break the MVC pattern

Server model is not passed to the constructor anymore, as well as the
Organisation model.

mokaddem committed:
chg: [EventController] replaced if/else by ternary condition

iglocska commented:
Awesome sauce, thanks a lot!
